### ISSUE TITLE
Preventing Mods From Using Dangerous JDK Packages

### ISSUE THREAD

**Issue by sylv256** [QuiltMC/quilt-loader#359]
It's imperative that mods be prevented from using dangerous JDK packages and `sun` implementation details. For example, we don't want mods re-transforming classes at runtime via Instrumentation without a command line argument. This is dangerous behavior and a mod should never need to do something like this.

We should also block the AWT package since that causes issues on macOS.

**Comment by AlexIIL** [comment]
Personally I don't like this idea - most mods don't need (or want) to do dangerous things like this, but modding isn't only about writing clean, safe code - we're a mod loader for modifying a game after all.

> and sun implementation details.

Sodium uses `sun.misc.Unsafe` for performance. If we wanted to disallow access to JDK internals we'd either cause crashes with sodium at runtime, or leave users with slower performance. (And not disallowing `Unsafe` allows mods to do basically anything).

> For example, we don't want mods re-transforming classes at runtime via Instrumentation without a command line argument

I think https://openjdk.org/jeps/451 does what you want here? I don't think quilt-loader has much control over this.

I'm a bit confused as to why you've opened this though - how much did you use your own library that does this? (https://github.com/VulpixelMC/YummyQuiltHacks)

**Comment by rinOfTheStars** [comment]
> This is dangerous behavior and a mod should never need to do something like this.

As you should know considering you worked on YQH, that's outright _incorrect_: patch-based mass transformation of vanilla's code is really the only way to make jarmods for older versions into loader mods; see https://github.com/BTW-Community/Cursed-BTW/tree/main for a project that does _exactly that_, though it's using a fork of fabric loader instead of quilt loader rn for 1.5.2-related reasons.

Is it dangerous behavior? Yes. Does it still have legit use cases? Also yes. Therefor, QL shouldn't be blocking said behavior by default, especially when QSL/QFAPI isn't present—there's a case to restrict such behavior by default then (since it's guaranteed to be running on a modern version where mass class transformation is bound to be a lot less common), but loader itself shouldn't be making that decision.

**Comment by sylv256** [comment]
> I think https://openjdk.org/jeps/451 does what you want here? I don't think quilt-loader has much control over this.

Yeah, that's more of what I meant. Throwing an error to prevent it from happening easily would be nice though.

> I'm a bit confused as to why you've opened this though - how much did you use your own library that does this? (https://github.com/VulpixelMC/YummyQuiltHacks)

I don't see any reason to be self-attaching an agent for instrumentation. If the loader wants to implement instrumentation, it can, but that's up to the loader, not the modder. Additionally, self-attaching agents has the potential to break on some JVMs. I somewhat regret writing YQH the way I did, and if I did it again, it would likely be without instrumentation. Either the loader supports instrumentation natively, or it doesn't. Also, I would avoid touching Quilt's internals as much as I did.

**Comment by sylv256** [comment]
> > This is dangerous behavior and a mod should never need to do something like this.
> 
> As you should know considering you worked on YQH, that's outright _incorrect_: patch-based mass transformation of vanilla's code is really the only way to make jarmods for older versions into loader mods; see https://github.com/BTW-Community/Cursed-BTW/tree/main for a project that does _exactly that_, though it's using a fork of fabric loader instead of quilt loader rn for 1.5.2-related reasons.
> 
> Is it dangerous behavior? Yes. Does it still have legit use cases? Also yes. Therefor, QL shouldn't be blocking said behavior by default, especially when QSL/QFAPI isn't present—there's a case to restrict such behavior by default then (since it's guaranteed to be running on a modern version where mass class transformation is bound to be a lot less common), but loader itself shouldn't be making that decision.

Are you saying that it would be reasonable for QSL to block runtime re-transformation by default? I agree with that at least.

**Comment by rinOfTheStars** [comment]
IMO if such a feature is added (a mass transformation permissibility toggle) it would make the most sense as a loader config option with three settings: `ON`, `OFF`, and `DEFAULT`—QSL/QFAPI/FAPI if someone's running that on QL for whatever reason would cause `DEFAULT` to behave like 'OFF' (meaning mass transformation is disabled), while pure QL would have it behave like `ON` (mass transformation is allowed). Either that, or allowing for the default behavior to be overridden with a JVM argument?

**Comment by sylv256** [comment]
> 

I don't want to disallow mass transformation (via mixin plugins), I just want to disallow Instrumentation by default.

**Comment by kb-1000** [comment]
Why? For security, this is useless. JDK says that lock-down stuff is for implementation "integrity". I don't buy it, and "integrity" gets more questionable in modding too.
So what's the purpose of this? Enforcing mods to have "good" code, effectively forcing those this would affect to have worse or no code?

**Comment by sylv256** [comment]
> Why? For security, this is useless. JDK says that lock-down stuff is for implementation "integrity". I don't buy it, and "integrity" gets more questionable in modding too. So what's the purpose of this? Enforcing mods to have "good" code, effectively forcing those this would affect to have worse or no code?

Why in the world do you need Instrumentation without a command line argument? The JDK is going to replace this functionality by default soon anyways, so I don't see a point in allowing it. If anything, we should be preparing for it to be disabled by default.

**Comment by kb-1000** [comment]
requiring command line arguments means exactly one thing: forcing those that must to relaunch the java process, aka worse code (or those that don't want to to not make their mod, aka no code).
Locking down all sun.* is definitely completely unreasonable given the fact Unsafe is there. Instrumentation only you _somewhat_ have a point there but not really....
What's the purpose of this? Why do you care?

**Comment by sylv256** [comment]
> requiring command line arguments means exactly one thing: forcing those that must to relaunch the java process, aka worse code (or those that don't want to to not make their mod, aka no code). Locking down all sun.* is definitely completely unreasonable given the fact Unsafe is there. Instrumentation only you _somewhat_ have a point there but not really.... What's the purpose of this? Why do you care?

It's all about encouraging good coding practices. I feel like if anything were implemented, the Instrumentation mitigation would be. I don't want to limit Unsafe at all, but disabling or locking out the AWT package could be more beneficial.

**Comment by lukebemish** [comment]
Frankly, this is a terrible idea. It is not quilt loader's place to prevent modders from doing things that they might have completely legitimate reasons for having to do. Yes, use of AWT classes is generally a bad thing. No, quilt loader is not the person to say that that's always the case and someone isn't doing something funky where that makes sense. Same with instrumentation; there are legitimate reasons to be using it, and placing arbitrary restrictions beyond what Java already has is silly.

...and all of that said if you allow the use of Unsafe at all you can't prevent anyone from doing any of that other stuff anyways

**Comment by sylv256** [comment]
> Frankly, this is a terrible idea. It is not quilt loader's place to prevent modders from doing things that they might have completely legitimate reasons for having to do. Yes, use of AWT classes is generally a bad thing. No, quilt loader is not the person to say that that's always the case and someone isn't doing something funky where that makes sense. Same with instrumentation; there are legitimate reasons to be using it, and placing arbitrary restrictions beyond what Java already has is silly.
> 
> ...and all of that said if you allow the use of Unsafe at all you can't prevent anyone from doing any of that other stuff anyways

The point is not to prevent entirely but to discourage.

**Comment by lukebemish** [comment]
Yes but... Why? Why is that quilt loader's job? It seems to me like you're saying "people shouldn't do hacky stuff and quilt loader should enforce that" but... We're in the realm of modding. Doing hacky stuff is the bread and butter of modding.

Let me rephrase the question: what _actual issue_ would this solve? Where's the example of a mod that was using this stuff _when it didn't need to_ and breaking stuff that this would prevent?

**Comment by sylv256** [comment]
A great example is mods (like Emogg) using the AWT classes. This could easily be fixed by simply refusing to load the AWT classes in Knot.

I believe one of auoeke's mods uses fabric loader internals or something fabric-specific which breaks on Quilt.

Either way, these things are fragile, far more fragile than Mixin (an intended feature) with Minecraft is. Preventing access to Quilt Loader internals can also help since those mods would be loader-version-dependent using impl details. I think generally preventing (or having a feature to prevent) access to impl without hacky solutions would be a good feature and would save headaches (e.g. preventing mods from Mixin' into Sodium internals) and allowing for some way to get around it manually, more as a way to say "this is not supported and never will be. We warned you."

**Comment by AlexIIL** [comment]
You can use AWT classes just fine in quilt - minecraft itself still uses `BufferedImage` in `MinecraftServer` for example. It's only a problem if you use any of the non-headless stuff, as that can break on mac os.

**Comment by OroArmor** [comment]
Also quilt does limit access to loader internals now

**Comment by kb-1000** [comment]
Even using non-headless stuff can be done safely by special-casing macOS, as loader itself does (although valid use cases for that are rare with tinyfd existing for simple message boxes and file dialogs)

**Comment by rinOfTheStars** [comment]
Yeah, honestly, this doesn't really feel like something that should be QL's job, _especially_ if future JVM versions are going to make it harder to do anyways—if modders are going to do very unsafe things, they probably know that what they are doing isn't exactly best practice and are doing it anyways, and trying to make them Not Do That via QL is unlikely to be a real deterrent. 

There is an argument to be made, _maybe_, for a large, hard-to-miss dev environment only warning saying 'hey you're using dangerous features of the JVM that are very much not best practice in your project, don't use them unless you really know what you're doing' when launching and building projects containing said unsafe code, but actually restricting that behavior isn't really something QL should be doing.

**Comment by sylv256** [comment]
Going to close this then.

**Comment by SylvKT** [comment]
Closed for now.
